`Cannot find module` errors

The following error is happening while bundling Functions:

```
In file "/opt/build/repo/functions/buttondown.js": Cannot find module 'isomorphic-unfetch/package.json' from '/opt/build/repo/functions' 
    node_modules/resolve/lib/async.js:121:35 
    node_modules/resolve/lib/async.js:274:39 processDirs
    node_modules/resolve/lib/async.js:281:32 isdir
    node_modules/resolve/lib/async.js:23:69 
    fs.js:167:21 FSReqCallback.oncomplete
```

--- a/package.json
+++ b/package.json
@@ -50,6 +50,7 @@
     "cp-file": "^7.0.0",
     "elf-tools": "^1.1.1",
     "end-of-stream": "^1.4.4",
+    "find-up": "^4.1.0",
     "glob": "^7.1.6",
     "make-dir": "^3.0.2",
     "p-map": "^3.0.0",
@@ -58,6 +59,7 @@
     "precinct": "^6.2.0",
     "require-package-name": "^2.0.1",
     "resolve": "^2.0.0-next.1",
+    "semver": "^7.3.2",
     "unixify": "^1.0.0",
     "util.promisify": "^1.0.1",
     "yargs": "^15.3.1"

--- a/src/dependencies.js
+++ b/src/dependencies.js
@@ -5,7 +5,7 @@ const precinct = require('precinct')
 const requirePackageName = require('require-package-name')
 const promisify = require('util.promisify')
 
-const { resolveLocation } = require('./resolve')
+const { resolvePathPreserveSymlinks, resolvePackage } = require('./resolve')
 
 const pGlob = promisify(glob)
 
@@ -64,7 +64,7 @@ const LOCAL_IMPORT_REGEXP = /^(\.|\/)/
 
 // When a file requires another one, we apply the top-level logic recursively
 const getLocalImportDependencies = async function(dependency, basedir, packageJson, state) {
-  const dependencyPath = await resolveLocation(dependency, basedir)
+  const dependencyPath = await resolvePathPreserveSymlinks(dependency, basedir)
   const depsPath = await getFileDependencies(dependencyPath, packageJson, state)
   return [dependencyPath, ...depsPath]
 }
@@ -89,7 +89,7 @@ const getModuleNameDependencies = async function(moduleName, basedir, state) {
   }
 
   // Find the Node.js module directory path
-  const packagePath = await resolveLocation(`${moduleName}/package.json`, basedir)
+  const packagePath = await resolvePackage(moduleName, basedir)
   const modulePath = dirname(packagePath)
 
   if (state.modulePaths.includes(modulePath)) {

--- a/src/resolve.js
+++ b/src/resolve.js
@@ -1,12 +1,54 @@
-const resolve = require('resolve')
+const { version: nodeVersion } = require('process')
 
-// Like `require.resolve()` but works with a custom base directory.
+const findUp = require('find-up')
+const pathExists = require('path-exists')
+const resolve = require('resolve')
+const { lt: ltVersion } = require('semver')
+
+// Find the path to a module's `package.json`
+// We need to use `resolve` instead of `require.resolve()` because:
+//  - it is async
+//  - it preserves symlinks:
+//     - this is important because if a file does a `require('./symlink')`, we
+//       need to bundle the symlink and its target, not only the target
+//     - `path.resolve()` cannot be used for relative|absolute file paths
+//       because it does not resolve ommitted file extension,
+//       e.g. `require('./file')` instead of `require('./file.js')`
+//     - the CLI flag `--preserve-symlinks` can be used with Node.js, but it
+//       cannot be set runtime
+// However it does not give helpful error messages.
+//   https://github.com/browserify/resolve/issues/223
+// So, on errors, we fallback to `require.resolve()`
+const resolvePackage = async function(moduleName, basedir) {
+  try {
+    return await resolvePathPreserveSymlinks(`${moduleName}/package.json`, basedir)
+  } catch (error) {
+    if (ltVersion(nodeVersion, REQUEST_RESOLVE_MIN_VERSION)) {
+      throw error
+    }
+
+    try {
+      return resolvePathFollowSymlinks(`${moduleName}/package.json`, basedir)
+    } catch (error) {
+      const packagePath = resolvePackageFallback(moduleName, basedir)
+      if (packagePath === undefined) {
+        throw error
+      }
+      return packagePath
+    }
+  }
+}
+
+// TODO: remove after dropping support for Node <8.9.0
+// `require.resolve()` option `paths` was introduced in Node 8.9.0
+const REQUEST_RESOLVE_MIN_VERSION = '8.9.0'
+
 // We need to use `new Promise()` due to a bug with `utils.promisify()` on
 // `resolve`:
 //   https://github.com/browserify/resolve/issues/151#issuecomment-368210310
-const resolveLocation = function(location, basedir) {
+const resolvePathPreserveSymlinks = function(path, basedir) {
   return new Promise((success, reject) => {
-    resolve(location, { basedir, preserveSymlinks: true }, (error, resolvedLocation) => {
+    resolve(path, { basedir, preserveSymlinks: true }, (error, resolvedLocation) => {
       if (error) {
         return reject(error)
       }
@@ -16,4 +58,33 @@ const resolveLocation = function(location, basedir) {
   })
 }
 
-module.exports = { resolveLocation }
+const resolvePathFollowSymlinks = function(path, basedir) {
+  return require.resolve(path, { paths: [basedir] })
+}
+
+// `require.resolve()` on a module's specific file (like `package.json`)
+// can be forbidden by the package author by using an `exports` field in
+// their `package.json`. We need this fallback.
+// It looks for the first directory up from a package's `main` file that:
+//   - is named like the package
+//   - has a `package.json`
+// Theoritically, this might not the root `package.json`, but this is very
+// unlikely, and we don't have any better alternative.
+const resolvePackageFallback = async function(moduleName, basedir) {
+  const mainFilePath = resolvePathFollowSymlinks(moduleName, basedir)
+  return findUp(isPackageDir.bind(null, moduleName), { cwd: mainFilePath, type: 'directory' })
+}
+
+const isPackageDir = async function(moduleName, dir) {
+  // Need to use `endsWith()` to take into account `@scope/package`.
+  // Backslashes need to be converted for Windows.
+  if (!dir.replace(BACKSLASH_REGEXP, '/').endsWith(moduleName) || !(await pathExists(`${dir}/package.json`))) {
+    return
+  }
+
+  return dir
+}
+
+const BACKSLASH_REGEXP = /\\/g
+
+module.exports = { resolvePackage, resolvePathPreserveSymlinks }

